executor/aggfuncs/func_group_concat_test.go

```diff
@@ -45,3 +52,272 @@ func (s *testSuite) TestGroupConcat(c *C) {
+	memDeltas = make([]int64, 0)
+	buffer := new(bytes.Buffer)
+	dataGen := getDataGenFunc(dataType)
+	for i := 0; i < srcChk.NumRows(); i++ {
+		row := srcChk.GetRow(i)
+		if row.IsNull(0) {
+			memDeltas = append(memDeltas, int64(0))
+			continue
+		}
+		oldMemSize := buffer.Len()
+		if i != 0 {
+			buffer.WriteString(" ")
+		}
+		curVal := row.GetString(0)
+		buffer.WriteString(curVal)
+		dt := dataGen(i)
+		datumMem := aggfuncs.GetDatumMemSize(&dt)
+		memDelta := datumMem + int64(buffer.Len()-oldMemSize)
+		memDeltas = append(memDeltas, memDelta)
+	}
+	return memDeltas, nil
+}
+
+func groupConcatDistinctUpdateMemDeltaGens(srcChk *chunk.Chunk, dataType *types.FieldType) (memDeltas []int64, err error) {
+	valSet := set.NewStringSet()
+	buffer := new(bytes.Buffer)
+	var encodeBytesBuffer []byte
+	for i := 0; i < srcChk.NumRows(); i++ {
+		row := srcChk.GetRow(i)
+		if row.IsNull(0) {
+			memDeltas = append(memDeltas, int64(0))
+			continue
+		}
+		encodeBytesBuffer = encodeBytesBuffer[:0]
+		curVal := row.GetString(0)
+		encodeBytesBuffer = codec.EncodeBytes(encodeBytesBuffer, hack.Slice(curVal))
+		joinedVal := string(encodeBytesBuffer)
+		if valSet.Exist(joinedVal) {
+			memDeltas = append(memDeltas, int64(0))
+			continue
+		}
+		oldMemSize := buffer.Cap()
+		if i != 0 {
+			buffer.WriteString(" ")
+		}
+		valSet.Insert(joinedVal)
+		buffer.WriteString(curVal)
+		memDelta := int64(len(joinedVal) + (buffer.Cap() - oldMemSize))
+		memDeltas = append(memDeltas, memDelta)
+	}
+	return memDeltas, nil
+}
+
+func groupConcatDistinctOrderUpdateMemDeltaGens(srcChk *chunk.Chunk, dataType *types.FieldType) (memDeltas []int64, err error) {
+	valSet := set.NewStringSet()
+	buffer := new(bytes.Buffer)
+	var encodeBytesBuffer []byte
+	dataGen := getDataGenFunc(dataType)
+	for i := 0; i < srcChk.NumRows(); i++ {
+		row := srcChk.GetRow(i)
+		if row.IsNull(0) {
+			memDeltas = append(memDeltas, int64(0))
+			continue
+		}
+		encodeBytesBuffer = encodeBytesBuffer[:0]
+		curVal := row.GetString(0)
+		encodeBytesBuffer = codec.EncodeBytes(encodeBytesBuffer, hack.Slice(curVal))
+		joinedVal := string(encodeBytesBuffer)
+		if valSet.Exist(joinedVal) {
+			memDeltas = append(memDeltas, int64(0))
+			continue
+		}
+		oldMemSize := buffer.Len()
+		if i != 0 {
+			buffer.WriteString(" ")
+		}
+		valSet.Insert(joinedVal)
+		buffer.WriteString(curVal)
+		dt := dataGen(i)
+		datumMem := aggfuncs.GetDatumMemSize(&dt)
+		memDelta := datumMem + int64(len(joinedVal)+(buffer.Len()-oldMemSize))
+		memDeltas = append(memDeltas, memDelta)
+	}
+	return memDeltas, nil
+}
+
+func groupConcatMultiArgsUpdateMemDeltaGens(srcChk *chunk.Chunk, dataType []*types.FieldType, byItems []*util.ByItems) (memDeltas []int64, err error) {
+	memDeltas = make([]int64, 0)
+	buffer := new(bytes.Buffer)
+	for i := 0; i < srcChk.NumRows(); i++ {
+		row := srcChk.GetRow(i)
+		if row.IsNull(0) {
```

Comment:
there are multi args, we need to check `IsNull` for every args

Reply:
It makes me confuse that why `srcChk.AppendDatum(0, &types.Datum{})` must be necessary？If not, we don't need to check `IsNull`.  In other words,  `row.IsNull(0)` check just for `srcChk.AppendDatum(0, &types.Datum{})`, no matter multi args or not. Is it true?
```
func (p *multiArgsAggTest) genSrcChk() *chunk.Chunk {
	srcChk := chunk.NewChunkWithCapacity(p.dataTypes, p.numRows)
	for i := 0; i < p.numRows; i++ {
		for j := 0; j < len(p.dataGens); j++ {
			fdt := p.dataGens[j](i)
			srcChk.AppendDatum(j, &fdt)
		}
	}
	srcChk.AppendDatum(0, &types.Datum{})
	return srcChk
}
```

Reply:
It's ok in the current situation, but I think every column should be checked more rigorously.